src/geoip_cache.rs

```diff
@@ -0,0 +1,105 @@
+
+#[derive(Clone, Serialize, Deserialize)]
+struct CacheEntry {
+    pub last_updated: SystemTime,
+    pub entry: GeoIPInfo,
+}
+
+#[derive(Default, Clone, Serialize, Deserialize)]
+struct GeoCache {
+    pub entries: HashMap<IpAddr, CacheEntry>,
+}
+
+/// GeoIP cache responsible for getting and caching results.
+pub struct GeoIPCache {
+    /// Available providers and their configuration.
+    providers: Vec<Box<dyn GeoIPService>>,
+    /// Path to the cache file.
+    cache_file: PathBuf,
+    /// Cache entries.
+    cache: Arc<RwLock<GeoCache>>,
+}
+
+impl GeoIPCache {
+    /// Create a new GeoIP cache.
+    pub fn new(cache_file: &Path) -> Self {
+        Self {
+            providers: Vec::new(),
+            cache_file: cache_file.to_owned(),
+            cache: Arc::new(RwLock::new(GeoCache::default())),
+        }
+    }
+
+    /// Add a new provider to the list of providers. The providers will be called in the order they
+    /// are added.
+    pub fn add_provider(&mut self, provider: Box<dyn GeoIPService>) {
+        self.providers.push(provider);
+    }
+
+    /// Load the cache from the file.
+    pub async fn load(&self) -> Result<(), io::Error> {
+        let cache_string = fs::read_to_string(&self.cache_file)?;
+
+        let mut cache = self.cache.write().await;
+        cache.entries = serde_json::from_str(&cache_string).unwrap();
+        Ok(())
+    }
+
+    /// Save the cache to the file.
+    pub async fn save(&self) -> Result<(), io::Error> {
+        let cache = self.cache.read().await;
+        let cache_string = serde_json::to_string(&cache.entries).unwrap();
+        fs::write(&self.cache_file, cache_string)
+    }
+
+    /// Function look in cache and if not found, it will call the providers to fetch new data and
+    /// store it into cache.
+    pub async fn lookup(&self, ip: IpAddr) -> Option<GeoIPInfo> {
+        let mut remove_entry = false;
+
+        {
+            let cache = self.cache.read().await;
+            let res = cache.entries.get(&ip);
+            if let Some(entry) = res {
+                // Check if the entry is not too old.
+                if entry.last_updated.elapsed().unwrap() < Duration::from_secs(60 * 60 * 24 * 14) {
+                    return Some(entry.entry.clone());
+                } else {
+                    remove_entry = true;
+                }
+            }
+        }
+
+        if remove_entry {
+            let mut rw_cache = self.cache.write().await;
+            rw_cache.entries.remove(&ip);
+        }
+
+        for provider in self.providers.iter() {
+            let entry = provider.lookup(ip).await;
+            if let Ok(entry) = entry {
+                let mut rw_cache = self.cache.write().await;
+                let ret_entry = entry.clone();
+                rw_cache.entries.insert(
+                    ip,
+                    CacheEntry {
+                        last_updated: SystemTime::now(),
+                        entry,
```

Comment:
wouldn't this work without an extra variable:
```suggestion
                rw_cache.entries.insert(
                    ip,
                    CacheEntry {
                        last_updated: SystemTime::now(),
                        entry.clone(),
```

Reply:
In fact you're right it could but not exactly that way - field name is needed as follows:
```
rw_cache.entries.insert(
                    ip,
                    CacheEntry {
                        last_updated: SystemTime::now(),
                        entry: entry.clone(),
                    },
                );
```

Otherwise you get:
```
error: expected one of `,`, `:`, or `}`, found `.`
  --> src/geoip_cache.rs:96:30
   |
94 |                     CacheEntry {
   |                     ---------- while parsing this struct
95 |                         last_updated: SystemTime::now(),
96 |                         entry.clone(),
   |                         -    ^ expected one of `,`, `:`, or `}`
```

Reply:
Done.